infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/SecretAsContainerResourceProvisioner.java

```diff
@@ -0,0 +1,179 @@
+    for (PodData pd : env.getPodsData().values()) {
+      if (!pd.getRole().equals(PodRole.DEPLOYMENT)) {
+        continue;
+      }
+      for (Container c : pd.getSpec().getContainers()) {
+        if (targetContainerName != null && !c.getName().equals(targetContainerName)) {
+          continue;
+        }
+        for (Entry<String, String> entry : secret.getData().entrySet()) {
+          final String mountEnvName = envName(secret, entry.getKey());
+          c.getEnv()
+              .add(
+                  new EnvVarBuilder()
+                      .withName(mountEnvName)
+                      .withValueFrom(
+                          new EnvVarSourceBuilder()
+                              .withSecretKeyRef(
+                                  new SecretKeySelectorBuilder()
+                                      .withName(secret.getMetadata().getName())
+                                      .withKey(entry.getKey())
+                                      .build())
+                              .build())
+                      .build());
+        }
+      }
+    }
+  }
+
+  private void mountAsFile(E env, Secret secret, String targetContainerName)
+      throws InfrastructureException {
+
+    final String mountPath = secret.getMetadata().getAnnotations().get("mountPath");
+    if (mountPath == null) {
+      throw new InfrastructureException(
+          format(
+              "Unable to mount secret '%s': it has to be mounted as file, but mountPath is not specified."
+                  + "Please make sure you specified 'mountPath' annotation of secret.'",
+              secret.getMetadata().getName()));
+    }
+
+    Volume volumeFromSecret =
+        new VolumeBuilder()
+            .withName(secret.getMetadata().getName())
+            .withSecret(
+                new SecretVolumeSourceBuilder()
+                    .withNewSecretName(secret.getMetadata().getName())
+                    .build())
+            .build();
+
+    for (PodData pd : env.getPodsData().values()) {
+      if (!pd.getRole().equals(PodRole.DEPLOYMENT)) {
+        continue;
+      }
+      pd.getSpec().getVolumes().add(volumeFromSecret);
+      for (Container c : pd.getSpec().getContainers()) {
+        if (targetContainerName != null && !c.getName().equals(targetContainerName)) {
+          continue;
+        }
+        c.getVolumeMounts()
+            .add(
+                new VolumeMountBuilder()
+                    .withName(volumeFromSecret.getName())
+                    .withMountPath(mountPath)
+                    .withReadOnly(true)
+                    .build());
+      }
+    }
+  }
+
+  private String envName(Secret secret, String key) throws InfrastructureException {
+    String mountEnvName;
+    if (secret.getData().size() == 1) {
+      try {
+        mountEnvName =
+            firstNonNull(
+                secret.getMetadata().getAnnotations().get("envName"),
+                secret.getMetadata().getAnnotations().get(key + ".envName"));
```

Comment:
This may overflow the 63 chars limit on annotation name (note that the "domain name" before the `/` doesn't count towards that limit - https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set).

Maybe we should do something like:

`org.eclipse.che/secret_1=<name-of-secret-data>`
`org.eclipse.che/secret-env_1=<name-of-env-var-to-mount-as>`

Where the matching would be made by the string after the `_` (and it doesn't even have to be a number).

Reply:
User will be just unable to create such secret. So i would like to leave it as is, 55 symbols is more than enough for data entry key, and that you proposed will be extremely hard to explain to user what he should write in his secrets, and, especially, correctly state that in exception message (i am not even sure i did get it right :) ).

Reply:
So what with that @metlos @sparkoo ? I am not a big fan of creating separate properties to decribe single thing. it's very hard to document and extremely hard to provide correct error message in case something is missing. 55 symbols (63 - `_envName` ) is more than enouth for data key, otherwise, user will just not be able to create such secret (or code does not creates it).

Reply:
I'm ok with explaining the 55 character limit in the docs.